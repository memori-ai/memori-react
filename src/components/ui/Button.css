.memori-button {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    padding: 4px 15px;
    border: 1px solid var(--memori-button-border-color, #d9d9d9);
    border-radius: var(--memori-button-radius, 5px);
    background: var(--memori-button-bg, #fff);
    box-shadow: var(--memori-button-box-shadow);
    color: var(--memori-button-text, #333);
    font-family: var(--memori-font-family);
    font-size: 14px;
    transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);
}
.memori-button--active {
    background: var(--memori-primary);
    color: var(--memori-primary-text);
}

.memori-button:hover {
    box-shadow: 0 2px 0 rgba(0, 0, 0, 0.02), 0 4px 8px rgba(0, 0, 0, 0.04);
    color: #333;
    cursor: pointer;
}

.memori-button:hover,
.memori-button:focus {
    border-color: var(--memori-primary);
    color: var(--memori-primary);
}

.memori-button:focus {
    outline-color: var(--memori-primary);
}

.memori-button:active {
    box-shadow: inset var(--memori-button-box-shadow);
}

.memori-button--padded {
    padding: var(--memori-button-padding);
}

.memori-button--block {
    display: block;
    width: 100%;
}

.memori-button--primary {
    background: var(--memori-primary);
    color: var(--memori-primary-text);
}

.memori-button--primary:hover,
.memori-button--primary:focus {
    color: var(--memori-primary-text);
}

.memori-button--primary:active {
    transform: scale(0.95);
}

.memori-button--outlined {
    border-color: var(--memori-primary);
    background: none;
    color: var(--memori-primary);
}

.memori-button--outlined:hover,
.memori-button--outlined:focus {
    border-color: var(--memori-primary);
    background-color: var(--memori-primary);
    color: var(--memori-primary-text);
}

.memori-button--outlined:active {
    transform: scale(0.95);
}

.memori-button--ghost {
    border: 1px solid transparent;
    background: none;
    box-shadow: none;
}

.memori-button--ghost:hover,
.memori-button--ghost:focus {
    border-color: transparent;
    background: rgba(0, 0, 0, 0.04);
    box-shadow: 0 2px 0 rgba(0, 0, 0, 0.02), 0 4px 8px rgba(0, 0, 0, 0.04);
    color: inherit;
}

.memori-button--ghost:focus {
    border: 1px solid var(--memori-primary);
}

.memori-button--square {
    border-radius: 0;
}

.memori-button--rounded {
    border-radius: var(--memori-button-radius);
}

.memori-button--danger {
    border-color: var(--memori-error-color);
    background: rgba(0, 0, 0, 0.02);
    color: var(--memori-error-color);
}

.memori-button--danger:hover,
.memori-button--danger:focus {
    border-color: var(--memori-error-color);
    background-color: var(--memori-error-color);
    color: #fff;
}

.memori-button--danger.memori-button--ghost {
    border-color: transparent;
    background: none;
    box-shadow: none;
    color: var(--memori-error-color);
}

.memori-button--danger.memori-button--ghost:hover,
.memori-button--danger.memori-button--ghost:focus {
    border-color: var(--memori-error-color);
    background: rgba(0, 0, 0, 0.02);
    box-shadow: 0 2px 0 rgba(0, 0, 0, 0.02), 0 4px 8px rgba(0, 0, 0, 0.04);
    color: var(--memori-error-color);
}

.memori-button--circle {
    padding: 0.75em;
    border-radius: 50%;
    aspect-ratio: 1/1;
}

.memori-button--icon {
    display: inline-flex;
    color: currentColor;
}

.memori-button--icon svg {
    width: 1.25em;
    height: 1.25em;
    fill: currentColor;
    font-size: 1em;
    line-height: 1;
}

.memori-button--icon svg[fill='none'] {
    fill: none;
}

.memori-button--with-icon .memori-button--icon {
    margin-right: 0.5em;
}

.memori-button[disabled] {
    background: var(--memori-button-disabled-bg, #ccc);
    color: var(--memori-button-disabled-color, #000);
    cursor: not-allowed;
    opacity: 0.5;
}
