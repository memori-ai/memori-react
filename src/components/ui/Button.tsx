import React, { FC } from 'react';
import cx from 'classnames';
import Loading from '../icons/Loading';

export interface Props {
  children?:
    | React.ReactChild
    | React.ReactChild[]
    | React.ReactNode
    | JSX.Element;
  primary?: boolean;
  outlined?: boolean;
  ghost?: boolean;
  padded?: boolean;
  block?: boolean;
  icon?: React.ReactNode;
  shape?: 'square' | 'rounded' | 'circle';
  danger?: boolean;
  loading?: boolean;
  disabled?: boolean;
  className?: string;
  title?: string;
  id?: string;
  isActive?: boolean;
  htmlType?: 'button' | 'submit' | 'reset';
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  onMouseDown?: (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => void;
  onMouseUp?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  onMouseLeave?: (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => void;
  onTouchStart?: (
    event: React.TouchEvent<HTMLButtonElement> | React.MouseEvent
  ) => void;
  onTouchEnd?: (
    event: React.TouchEvent<HTMLButtonElement> | React.MouseEvent
  ) => void;
}

const Button: FC<Props> = ({
  primary = false,
  outlined = false,
  ghost = false,
  padded = true,
  shape = 'rounded',
  danger = false,
  loading = false,
  disabled = false,
  block = false,
  icon,
  className,
  title,
  id,
  htmlType,
  onClick,
  onMouseDown,
  onMouseUp,
  onMouseLeave,
  onTouchStart,
  onTouchEnd,
  children,
  isActive,
}) => (
  <button
    id={id}
    type={htmlType}
    onClick={onClick}
    onMouseDown={onMouseDown}
    onMouseUp={onMouseUp}
    onMouseLeave={onMouseLeave}
    onTouchStart={onTouchStart}
    onTouchEnd={onTouchEnd}
    title={title}
    disabled={loading || disabled}
    className={cx(
      'memori-button',
      {
        'memori-button--primary': primary,
        'memori-button--outlined': outlined,
        'memori-button--ghost': ghost,
        'memori-button--square': shape === 'square',
        'memori-button--rounded': shape === 'rounded',
        'memori-button--circle': shape === 'circle',
        'memori-button--padded': padded,
        'memori-button--block': block,
        'memori-button--with-icon': (icon || loading) && children,
        'memori-button--icon-only': (icon || loading) && !children,
        'memori-button--danger': danger,
        'memori-button--loading': loading,
        'memori-button--active': isActive,
      },
      className
    )}
  >
    {icon && !loading && <span className="memori-button--icon">{icon}</span>}
    {loading && (
      <span className="memori-button--icon loading-icon">
        <Loading loading />
      </span>
    )}
    {children}
  </button>
);

export default Button;
