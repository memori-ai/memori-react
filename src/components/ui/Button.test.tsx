import React from 'react';
import { render } from '@testing-library/react';
import Button from './Button';
import Loading from '../icons/Loading';

it('renders empty Button unchanged', () => {
  const { container } = render(<Button />);
  expect(container).toMatchSnapshot();
});

it('renders Button unchanged', () => {
  const { container } = render(<Button>Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button primary unchanged', () => {
  const { container } = render(<Button primary>Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button outlined unchanged', () => {
  const { container } = render(<Button outlined>Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button padded unchanged', () => {
  const { container } = render(<Button padded>Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button block unchanged', () => {
  const { container } = render(<Button block>Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button with icon unchanged', () => {
  const { container } = render(<Button icon={<Loading />}>Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button with icon only unchanged', () => {
  const { container } = render(<Button icon={<Loading />} />);
  expect(container).toMatchSnapshot();
});

it('renders Button danger unchanged', () => {
  const { container } = render(<Button danger>Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button loading unchanged', () => {
  const { container } = render(<Button loading>Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button disabled unchanged', () => {
  const { container } = render(<Button disabled>Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button with title unchanged', () => {
  const { container } = render(<Button title="Button title">Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button square unchanged', () => {
  const { container } = render(<Button shape="square">Click me</Button>);
  expect(container).toMatchSnapshot();
});

it('renders Button circle unchanged', () => {
  const { container } = render(<Button shape="circle" icon={<Loading />} />);
  expect(container).toMatchSnapshot();
});
