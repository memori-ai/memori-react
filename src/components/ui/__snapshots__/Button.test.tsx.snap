// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Button block unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--rounded memori-button--padded memori-button--block"
  >
    Click me
  </button>
</div>
`;

exports[`renders Button circle unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--circle memori-button--padded memori-button--icon-only"
  >
    <span
      class="memori-button--icon"
    >
      <svg
        aria-hidden="true"
        class="memori-loading-icon"
        focusable="false"
        role="img"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`renders Button danger unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--rounded memori-button--padded memori-button--danger"
  >
    Click me
  </button>
</div>
`;

exports[`renders Button disabled unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--rounded memori-button--padded"
    disabled=""
  >
    Click me
  </button>
</div>
`;

exports[`renders Button loading unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--rounded memori-button--padded memori-button--with-icon memori-button--loading"
    disabled=""
  >
    <span
      class="memori-button--icon loading-icon"
    >
      <svg
        aria-hidden="true"
        class="memori-loading-icon"
        focusable="false"
        role="img"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
        />
      </svg>
    </span>
    Click me
  </button>
</div>
`;

exports[`renders Button outlined unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--outlined memori-button--rounded memori-button--padded"
  >
    Click me
  </button>
</div>
`;

exports[`renders Button padded unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--rounded memori-button--padded"
  >
    Click me
  </button>
</div>
`;

exports[`renders Button primary unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--primary memori-button--rounded memori-button--padded"
  >
    Click me
  </button>
</div>
`;

exports[`renders Button square unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--square memori-button--padded"
  >
    Click me
  </button>
</div>
`;

exports[`renders Button unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--rounded memori-button--padded"
  >
    Click me
  </button>
</div>
`;

exports[`renders Button with icon only unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--rounded memori-button--padded memori-button--icon-only"
  >
    <span
      class="memori-button--icon"
    >
      <svg
        aria-hidden="true"
        class="memori-loading-icon"
        focusable="false"
        role="img"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`renders Button with icon unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--rounded memori-button--padded memori-button--with-icon"
  >
    <span
      class="memori-button--icon"
    >
      <svg
        aria-hidden="true"
        class="memori-loading-icon"
        focusable="false"
        role="img"
        viewBox="0 0 1024 1024"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
        />
      </svg>
    </span>
    Click me
  </button>
</div>
`;

exports[`renders Button with title unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--rounded memori-button--padded"
    title="Button title"
  >
    Click me
  </button>
</div>
`;

exports[`renders empty Button unchanged 1`] = `
<div>
  <button
    class="memori-button memori-button--rounded memori-button--padded"
  />
</div>
`;
